activerecord/lib/active_record/secure_password.rb

```diff
@@ -37,15 +35,17 @@ module ClassMethods
       #   User.authenticate_by(email: "jdoe@example.com") # => ArgumentError
       #   User.authenticate_by(password: "abc123")        # => ArgumentError
       def authenticate_by(attributes)
-        passwords = attributes.select { |name, value| !has_attribute?(name) && has_attribute?("#{name}_digest") }
+        passwords, identifiers = attributes.to_h.partition do |name, value|
+          !has_attribute?(name) && has_attribute?("#{name}_digest")
+        end.map(&:to_h)
 
         raise ArgumentError, "One or more password arguments are required" if passwords.empty?
-        raise ArgumentError, "One or more finder arguments are required" if passwords.size == attributes.size
+        raise ArgumentError, "One or more finder arguments are required" if identifiers.empty?
 
-        if record = find_by(attributes.except(*passwords.keys))
+        if record = find_by(identifiers)
```

Comment:
@kaspth I actually *find* (hah!) `find_by(attributes.except(*passwords.keys))` to be more clear, but I will defer to your judgment vis-a-vis c32234c25b56d0b0bb417cb8eb64ed4ef3e66b64. :+1: